python/cuml/utils/input_utils.py

```diff
@@ -248,3 +248,95 @@ def order_to_str(order):
+    host arrays.
+    Acceptable input formats:
+    * Numpy array - returns a pointer to the original input
+    * cuDF Dataframe - returns a deep copy always
+    * cuDF Series - returns by reference or a deep copy depending on
+        `deepcopy`
+    * cuda array interface compliant array (like Cupy) - returns a
+        reference unless deepcopy=True
+    * numba device array - returns a reference unless deepcopy=True
+
+    Returns: namedtuple('host_array', 'array pointer n_rows n_cols dtype')
+
+    `host_array` is a new device array if the input was not a NumPy device
+        array. It is a reference to the input X if it was a NumPy host array
+    """
+
+    if convert_to_dtype:
+        X = convert_dtype(X, to_dtype=convert_to_dtype)
+        check_dtype = False
+
+    if isinstance(X, cudf.DataFrame):
+        dtype = np.dtype(X[X.columns[0]]._column.dtype)
+        if order == 'F':
+            X_m = X.as_gpu_matrix(order='F')
+        elif order == 'C':
+            X_m = cuml.utils.numba_utils.row_matrix(X)
+        X_m = X_m.copy_to_host()
+
+    elif (isinstance(X, cudf.Series)):
+        if X.null_count == 0:
+            X_m = X.to_array()
+        else:
+            raise ValueError('cuDF Series has missing (null) values.')
+
+    elif isinstance(X, np.ndarray):
+        X_m = np.array(X, order=order, copy=deepcopy)
+
+    elif cuda.is_cuda_array(X):
+        # Use cuda array interface to create a device array by reference
+        X_m = cuda.as_cuda_array(X)
+        X_m = np.array(X_m.copy_to_host(), order=order)
+
+    else:
+        msg = "X matrix format " + str(X.__class__) + " not supported"
+        raise TypeError(msg)
+
+    dtype = X_m.dtype
+
+    if check_dtype:
+        if isinstance(check_dtype, type):
+            if dtype != check_dtype:
+                del X_m
+                raise TypeError("Expected " + str(check_dtype) + "input but" +
+                                " got " + str(dtype) + " instead.")
+        elif isinstance(check_dtype, Collection):
+            if dtype not in check_dtype:
+                del X_m
+                raise TypeError("Expected input to be of type in " +
+                                str(check_dtype) + " but got " + str(dtype))
+
+    n_rows = X_m.shape[0]
+    if len(X_m.shape) > 1:
+        n_cols = X_m.shape[1]
+    else:
+        n_cols = 1
+
+    if check_cols:
+        if n_cols != check_cols:
+            raise ValueError("Expected " + str(check_cols) +
+                             " columns but got " + str(n_cols) +
+                             " columns.")
+
+    if check_rows:
+        if n_rows != check_rows:
+            raise ValueError("Expected " + str(check_rows) +
+                             " rows but got " + str(n_rows) +
+                             " rows.")
+
+    X_ptr = X_m.ctypes.data
+
+    result = namedtuple('host_array', 'array pointer n_rows n_cols dtype')
```

Comment:
Maybe this namedtuple should be defined somewhere outside of the function in this file so it's reusable and shared with input_to_device_array?